File: balance/src/main/java/br/com/accenture/wallet/balance/repository/BalanceRepository.java
package br.com.accenture.wallet.balance.repository;

import br.com.accenture.wallet.balance.domain.Balance;
import io.micronaut.configuration.hibernate.jpa.scope.CurrentSession;
import io.micronaut.spring.tx.annotation.Transactional;

import javax.inject.Singleton;
import javax.persistence.*;
import java.util.List;
import java.util.Optional;
import java.util.Date;

@Singleton
public class BalanceRepository {
    private final String hqlFindByAccount = "SELECT b FROM Balance b WHERE b.accountId = :account";

    @PersistenceContext
    private EntityManager manager;

    public BalanceRepository(@CurrentSession EntityManager manager) {
        this.manager = manager;
    }

    @Transactional(readOnly = true)
    public Optional<Balance> findByAccount(String value) {
        TypedQuery<Balance> query = manager.createQuery(hqlFindByAccount, Balance.class);
        try {
            query.setParameter("account", value);
            return Optional.of(query.getSingleResult());
        } catch (NoResultException|NonUniqueResultException ex) {
            return Optional.empty();
        }
    }

    @Transactional(readOnly = true)
    public List<Balance> findTop5() {
        TypedQuery<Balance> query = manager.createQuery("SELECT b FROM Balance b ORDER BY b.accountId", Balance.class)
            .setMaxResults(5);
        return query.getResultList();
    }

    @Transactional
    public Optional<Balance> save(Balance bean) {
        try {
            manager.persist(bean);
            return Optional.of(bean);
        } catch (EntityExistsException ex) {
            return Optional.empty();
        }
    }

    @Transactional
    public Optional<Balance> update(Balance bean) {
        try {
            bean.setLastUpdate(new Date());
            manager.merge(bean);
            return Optional.of(bean);
        } catch (EntityExistsException ex) {
            return Optional.empty();
        }
    }

}

File: balance/src/main/kotlin/br/com/accenture/wallet/balance/repository/BalanceRepository.kt
package br.com.accenture.wallet.balance.repository

import br.com.accenture.wallet.balance.domain.Balance
import io.micronaut.configuration.hibernate.jpa.scope.CurrentSession
import io.micronaut.spring.tx.annotation.Transactional
import java.util.*
import javax.inject.Named
import javax.inject.Singleton
import javax.persistence.*

@Named
@Singleton
// open class must be declared because Hibernate proxy
open class BalanceRepository(@CurrentSession @PersistenceContext var manager: EntityManager) {

    private val findByAccount = "SELECT b FROM Balance b WHERE b.account = :account"

    @Transactional(readOnly = true)
    open fun findByAccount(value: String): Optional<Balance> {
        val query = manager.createQuery(findByAccount, Balance::class.java)
        return try {
            query.setParameter("account", value)
            Optional.of(query.singleResult)
        } catch (ex: NoResultException) {
            Optional.empty()
        } catch (ex: NonUniqueResultException) {
            Optional.empty()
        }
    }

    @Transactional
    open fun save(bean: Balance): Optional<Balance> {
        return try {
            manager.persist(bean)
            Optional.of(bean)
        } catch (ex: EntityExistsException) {
            Optional.empty()
        }
    }
}